### PR TITLE
Feature/fragment processing failure handling

### PR DESCRIPTION
**Pull request by scypio**
IMPLEMENTATION IN PROGRESS

## Description
Implementation of #466 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the [code style](/CONTRIBUTING.md#coding-conventions) of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

---
I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: knotx-core/src/main/java/io/knotx/dataobjects/Fragment.java [hunk 1]
```diff
@@ -36,38 +36,56 @@
...
+  private static final String FAILED_KEY = "_FAILED";
+  private static final String FAILURE_DETAILS_KEY = "_FAILEURE_DETAILS";
+  private static final String FALLBACK_KEY = "_FALLBACK";
+
 
   private final List<String> knots;
   private final JsonObject context;
   private String content;
+  private boolean failed;
+  private String failureDetails;
```
Comment (tomaszmichalak) [inline knotx-core/src/main/java/io/knotx/dataobjects/Fragment.java]: I think about the solution that instead of `failureDetails` can provide more details about the flow. We could extend `knots` list with such details as:
- knotIdentifier
- processed: SUCCESS, FAILURE, UNPROCESSED
- errors:
   - error code
   - error message

So if for example Data Bridge would not be able to call 2 services we  would get:
- knotIndetifier: dataBridge
- processed: FAILURE
- errors [ {error_from_service_1}, {error_from_service_2} ]

Comment (scypio) [reply knotx-core/src/main/java/io/knotx/dataobjects/Fragment.java]: changes implemented here, unfortunately github is not working that great right now so they are not visible in the PR: https://github.com/scypio/knotx/commit/66be8f930a7a15140ff21c6bae31e2dcb33dc47d

Comment (scypio) [reply knotx-core/src/main/java/io/knotx/dataobjects/Fragment.java]: added Knot object to Fragment with knot name / status and errors information

File: knotx-core/src/main/java/io/knotx/dataobjects/Fragment.java [hunk 2]
```diff
@@ -36,44 +37,66 @@
...
   private static final String CONTEXT_KEY = "_CONTEXT";
+  private static final String FAILED_KEY = "_FAILED";
+  private static final String FALLBACK_KEY = "_FALLBACK";
 
-  private final List<String> knots;
+
+  private final List<Knot> knots;
   private final JsonObject context;
   private String content;
+  private boolean failed;
```
[the remark this answers is not in the document]
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/dataobjects/Fragment.java]: removed

File: knotx-core/src/main/java/io/knotx/dataobjects/Fragment.java [hunk 3]
```diff
@@ -102,7 +125,31 @@ public JsonObject context() {
...
+   */
+  public boolean failed() {
+    return failed;
+  }
+
+  /**
+   * @return replacement markup that should be rendered if this Fragment has failed. Can be empty
+   * string. Null value indicates that no replacement markup is provided.
+   */
+  public String fallback() {
```
[the remark this answers is not in the document]
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/dataobjects/Fragment.java]: changed

File: knotx-core/src/main/java/io/knotx/dataobjects/Fragment.java [hunk 4]
```diff
@@ -102,7 +125,31 @@ public JsonObject context() {
...
+    return fallback;
+  }
+
+  public Fragment failed(boolean failed) {
+    this.failed = failed;
```
[the remark this answers is not in the document]
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/dataobjects/Fragment.java]: adjusted

File: knotx-core/src/main/java/io/knotx/dataobjects/Knot.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+import com.google.common.collect.Maps;
+import io.vertx.codegen.annotations.DataObject;
+import io.vertx.core.json.JsonObject;
+import java.util.Map;
+
+@DataObject(inheritConverter = true)
+public class Knot {
+  private String name;
+  private KnotStatus status = KnotStatus.UNPROCESSED; // SUCCESS, FAILURE, UNPROCESSED
+  private Map<String, Object> errors = Maps.newHashMap(); // key: errorCode, value: errorMessage
```
[the remark this answers is not in the document]
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/dataobjects/Knot.java]: class introduced, converted to list

File: knotx-core/src/main/java/io/knotx/fragments/SnippetPatterns.java [hunk 6]
```diff
@@ -26,9 +26,14 @@
   private static final String SNIPPET_PATTERN =
       "<%s\\s+%s" + FragmentConstants.SNIPPET_IDENTIFIER_NAME
           + "\\s*=\\s*\"([A-Za-z0-9-,]+)\"[^>]*>.+?</%s>";
+  private static final String SNIPPET_WITH_FALLBACK_PATTERN =
+      "<%s\\s+%s" + FragmentConstants.SNIPPET_IDENTIFIER_NAME
+          + "\\s*=\\s*\"([A-Za-z0-9-,]+)\"[^>]*"
```
[the remark this answers is not in the document]
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/fragments/SnippetPatterns.java]: we have unit tests and they work, but this part will be rewritten anyway

File: knotx-core/src/main/java/io/knotx/knot/AbstractKnotProxy.java [hunk 7]
```diff
@@ -72,6 +73,16 @@ public void process(KnotContext knotContext, Handler<AsyncResult<KnotContext>> r
   /**
    * Method lets you decide whether the Fragment should be processed by your Knot or not.
    *
+   * @param fragment Fragment to process - or skip  {@link Fragment}.
+   * @return <tt>true</tt> if this Knot should process current {@link Fragment}.
+   */
+  protected boolean shouldProcess(Fragment fragment) {
```
[the remark this answers is not in the document]
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/knot/AbstractKnotProxy.java]: That is a tricky one, shouldProcess was accepting only Set<Knot> beforehand. 
`protected abstract Single<KnotContext> processRequest(KnotContext knotContext);`

We should mark that method as deprecated and go with 
```
protected boolean shouldProcess(Set<String> knots, Fragment fragment) {
  // default implementation
}
```

File: knotx-core/src/main/java/io/knotx/util/FragmentUtil.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.knotx.util;
+
+import io.knotx.dataobjects.Fragment;
+import io.knotx.dataobjects.KnotStatus;
+
+public final class FragmentUtil {
+  public static Fragment failure(Fragment fragment, String knot, Throwable t) {
```
[the remark this answers is not in the document]
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/util/FragmentUtil.java]: moved to Fragment

File: knotx-core/src/main/java/io/knotx/util/FragmentUtil.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        .knotRouting().stream()
+        .filter(k -> knot.equals(k.getName()))
+        .findFirst()
+        .get()
+        .setStatus(KnotStatus.FAILURE)
+        .addError(t.getClass().getSimpleName(), t.getMessage());
+    return fragment;
+  }
+
+  public static Fragment success(Fragment fragment, String knot) {
```
[the remark this answers is not in the document]
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/util/FragmentUtil.java]: moved to Fragment

File: knotx-core/src/main/java/io/knotx/dataobjects/Fragment.java [hunk 10]
```diff
@@ -36,44 +38,62 @@
...
+  private final List<Knot> knots;
   private final JsonObject context;
   private String content;
+  private String fallback;
 
   public Fragment(JsonObject fragment) {
-    this.knots = fragment.getJsonArray(KNOTS_KEY).stream().map(String::valueOf)
-        .collect(Collectors.toList());
+    JsonArray knotsArray = fragment.getJsonArray(KNOTS_KEY);
+    this.knots = new ArrayList<>();
```
Comment (Skejven) [inline knotx-core/src/main/java/io/knotx/dataobjects/Fragment.java]: How about:
```suggestion
    this.knots= knotsArray.stream()
        .map(entry -> new Knot((JsonObject) entry))
        .collect(Collectors.toList());
```

If  you don't like doing this streams way, at least please initialize list with size `new ArrayList<>(knotsArray.size());`

File: knotx-core/src/main/java/io/knotx/fragments/SnippetPatterns.java [hunk 11]
```diff
@@ -26,9 +26,14 @@
   private static final String SNIPPET_PATTERN =
       "<%s\\s+%s" + FragmentConstants.SNIPPET_IDENTIFIER_NAME
           + "\\s*=\\s*\"([A-Za-z0-9-,]+)\"[^>]*>.+?</%s>";
+  private static final String SNIPPET_WITH_FALLBACK_PATTERN =
```
[the remark this answers is not in the document]
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/fragments/SnippetPatterns.java]: snippet implemented as suggested
```
<knotx:fallback data-knotx-fallback-id="MY_ID" data-knotx-fallback-strategy="MY_STRATEGY">
  any markup
</knotx:fallback>
```

File: knotx-core/src/main/java/io/knotx/knot/AbstractKnotProxy.java [hunk 12]
```diff
@@ -72,6 +73,16 @@ public void process(KnotContext knotContext, Handler<AsyncResult<KnotContext>> r
...
+    return fragment != null && !fragment.failed() && shouldProcess(Sets.newHashSet(fragment.knots()));
```
[the remark this answers is not in the document]
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/knot/AbstractKnotProxy.java]: redundant check removed

File: knotx-core/src/main/java/io/knotx/dataobjects/Fragment.java (hunk 10 above)
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/dataobjects/Fragment.java]: I like it, changed.

File: knotx-core/src/main/java/io/knotx/assembler/FragmentAssemblerKnotProxyImpl.java [hunk 13]
```diff
@@ -62,6 +63,11 @@ public FragmentAssemblerKnotProxyImpl(FragmentAssemblerOptions options) {
     }
   }
 
+  private String processFragment(Fragment fragment) {
+    return fragment.failed() && fragment.hasFallback() ? fragment.fallback()
```
[the remark this answers is not in the document]
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/assembler/FragmentAssemblerKnotProxyImpl.java]: Service loader implemented

File: knotx-core/src/main/java/io/knotx/dataobjects/Knot.java [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+
+import com.google.common.base.Objects;
+import com.google.common.collect.Lists;
+import io.vertx.codegen.annotations.DataObject;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+import java.util.List;
+
+@DataObject(inheritConverter = true)
+public class Knot {
```
Comment (Skejven) [inline knotx-core/src/main/java/io/knotx/dataobjects/Knot.java]: Correct me if I understood this wrong, but this class looks to me more like `KnotTask` or `KnotJob` (or smth. like that).
This is not `Knot` and naming it like that will introduce a lot of confusion. See here: https://github.com/Cognifide/knotx/wiki/Knot what `Knot` is.

If you will apply rename change here, please also think about `KnotStatus` and `KnotError` classes.

File: knotx-core/src/main/java/io/knotx/fragments/SnippetPatterns.java [hunk 15]
```diff
@@ -38,13 +45,24 @@ public SnippetPatterns(SnippetOptions snippetOptions) {
...
+                snippetOptions.getParamsPrefix(), snippetOptions.getTagName()), Pattern.DOTALL);
+    fallbackPattern = Pattern
+        .compile(String
+            .format(FALLBACK_PATTERN, snippetOptions.getFallbackTagName(),snippetOptions.getFallbackTagName()), Pattern.DOTALL);
   }
 
-  public Pattern getAnySnippetPattern() {
-    return anySnippetPattern;
-  }
+  public Pattern getAnySnippetPattern() { return anySnippetPattern; }
```
Comment (Skejven) [inline knotx-core/src/main/java/io/knotx/fragments/SnippetPatterns.java]: Please make it 3-liner again (as the methods below).
Now it is inconsistent with google formatting rules for java that we follow (see more in [contributing guide](https://github.com/Cognifide/knotx/blob/master/CONTRIBUTING.md#coding-conventions))

File: knotx-core/src/main/java/io/knotx/splitter/HtmlFragmentSplitter.java [hunk 16]
```diff
@@ -35,14 +37,17 @@
 
   @Override
   public List<Fragment> split(String html) {
+    if (StringUtils.isEmpty(html)) {
+      throw new NoSuchElementException("html cannot be empty");
+    }
     List<Fragment> fragments = Lists.newLinkedList();
```
Comment (Skejven) [inline knotx-core/src/main/java/io/knotx/splitter/HtmlFragmentSplitter.java]: I know that this is not part of the change you introduced, but could you change this to `ArrayList` here?
Thank you.

File: knotx-core/src/main/java/io/knotx/splitter/HtmlFragmentSplitter.java [hunk 17]
```diff
@@ -51,19 +56,41 @@
...
     } else {
-      fragments.add(toRaw(html, 0, html.length()));
+      processMarkup(fragments, html, 0, html.length());
     }
     return fragments;
   }
 
-  private Fragment toRaw(String html, int startIdx, int endIdx) {
-    return Fragment.raw(html.substring(startIdx, endIdx));
+  private void processMarkup(List<Fragment> fragments, String data, int startIdx, int endIdx) {
```
Comment (Skejven) [inline knotx-core/src/main/java/io/knotx/splitter/HtmlFragmentSplitter.java]: Now, for each found fragment, searching for fallbacks is triggered and searching for matching fallback is done from the beginning. Wouldn't be more efficient, to build some cache/map of fallbacks first and then only add proper fallback (if exists) by name?

File: knotx-core/src/main/java/io/knotx/dataobjects/Knot.java (hunk 14 above)
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/dataobjects/Knot.java]: renamed to KnotTask

File: knotx-core/src/main/java/io/knotx/fragments/SnippetPatterns.java (hunk 15 above)
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/fragments/SnippetPatterns.java]: filer reformatted

File: knotx-core/src/main/java/io/knotx/splitter/HtmlFragmentSplitter.java (hunk 16 above)
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/splitter/HtmlFragmentSplitter.java]: done

File: knotx-core/src/main/java/io/knotx/splitter/HtmlFragmentSplitter.java (hunk 17 above)
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/splitter/HtmlFragmentSplitter.java]: Introduced FallbackMarkers to run the fallback matcher once against the whole html

**Comment by scypio** [comment]
Features implemented so far:

### 1. Simple blank fallback
just add `knotx-data-fallback="BLANK"` to your snippet:
```
<knotx-snippet data-knotx-knots="databridge,handlebars" 
  data-knotx-databridge-name="unstable-service"
  data-knotx-fallback="BLANK">
      {{#if _result}}<h2>{{_result.count}}</h2>{{/if}}
</knotx-snippet>
```
### 2. Static markup fallback
create a fallback snippet with `data-knotx-fallback-id` attribute and point to it in your snippet:
```
<knotx-snippet data-knotx-knots="databridge,handlebars" 
  data-knotx-databridge-name="unstable-service"
  data-knotx-fallback="CUSTOM">
      {{#if _result}}<h2>{{_result.count}}</h2>{{/if}}
</knotx-snippet>
<knotx:fallback data-knotx-fallback-id="CUSTOM">
  <p class="error">error</p>
</knotx:fallback>
```
### 3. Global blank fallback
to apply BLANK fallback to all snippets by default configure `defaultFallback` attribute within `snippetOptions` - apply these to both Splitter and FragmentAssembler. 
```
  snippetOptions {
    tagName = knotx-snippet
    paramsPrefix = data-knotx-
    defaultFallback = BLANK
    ]
  }
```
### 4. Global fallback with custom markup
to apply custom markup fallback to all snippets by default: 
- configure `defaultFallback` attribute within `snippetOptions` 
- define your own `fallbacks` property within `snippetOptions`
- apply these to both Splitter and FragmentAssembler. 
```
  snippetOptions {
    tagName = knotx-snippet
    paramsPrefix = data-knotx-
    defaultFallback = CUSTOM_GLOBAL
    fallbacks = [
      {
        id = CUSTOM_GLOBAL
        markup = "<knotx:fallback data-knotx-fallback-id='CUSTOM_GLOBAL'><p class="error">error</p></knotx:fallback>"
      }
    ]
  }
```
### 5. fallback with custom FallbackStrategy
you can deliver your own class that will be fired by the assembler if given fragment fails
- implement FallbackStrategy interface
```
import io.knotx.dataobjects.Fragment;
import io.knotx.dataobjects.KnotContext;
import io.knotx.fallback.FallbackStrategy;

public class MyFallbackStrategy implements FallbackStrategy {

  @Override
  public String getId() {
    return "CUSTOM_STRATEGY";
  }

  @Override
  public String applyFallback(Fragment failed, Fragment fallback, KnotContext knotContext) {
    return "<p>markup</p>";
  }
}
```
- make sure that `ServiceLoader` can discover this class - add its fully qualified name to `/META-INF/services/io.knotx.fallback.FallbackStrategy` file. 
- create a custom fallback snippet with linked strategy Id
```
<knotx-snippet data-knotx-knots="databridge,handlebars" 
  data-knotx-databridge-name="unstable-service"
  data-knotx-fallback="CUSTOM">
      {{#if _result}}<h2>{{_result.count}}</h2>{{/if}}
</knotx-snippet>
<knotx:fallback data-knotx-fallback-id="CUSTOM" data-knotx-fallback-strategy="CUSTOM_STRATEGY">
  <p class="error">error</p>
</knotx:fallback>
```

**Comment by tomaszmichalak** [comment]
I am wondering if this tag `data-knotx-fallback="BLANK"` is really required. We can assume that if there is no `data-knotx-fallback-id` tag defined, then the default fallback mechanism should handle the snippet.

**Comment by tomaszmichalak** [comment]
According to 5:
We agreed that a fallback strategy should be configured as a chain, something like:
```
"fallbackStrategies": [
   { "name": "CUSTOM", conf":"{next:"DEFAULT"" }},
   { "name": "DEFAULT", "conf":"EMPTY" }
]
```
In that way our fallback strategy implementation should pass the flow to the next fallback. I thought about something similar to io.knotx.server.handler.knot.KnotEngineRoutingHandlerFactory.KnotxEngineHandler#handleRoute.

**Comment by tomaszmichalak** [comment]
Can we move this comment to the `SnippetFallback.md` file - we could work on documentation in the same PR. We plan to move documentation from wiki to *.md files so it can be good option.

File: knotx-core/src/main/java/io/knotx/splitter/HtmlFragmentSplitter.java [hunk 18]
```diff
@@ -51,19 +60,62 @@
...
+    int idx = startIdx;
+    for (FallbackMarker fe : fallbackMarkers) {
+      if (idx > fe.start) {
+        continue;
+      } else if (fe.start > endIdx) {
+        break;
+      } else if (idx < fe.start) {
+        fragments.add(Fragment.raw(html.substring(idx, fe.start)));
+      }
+      fragments.add(toFallback(html, fe.start, fe.end));
```
Comment (Skejven) [inline knotx-core/src/main/java/io/knotx/splitter/HtmlFragmentSplitter.java]: Maybe it would be worth to keep additional field in the `Fragment` e.g. `fragmentId` and extract it here?
That will allow to not use markup processing (JSoup) at the Assembler level to extract fallback id.

Comment (scypio) [reply knotx-core/src/main/java/io/knotx/splitter/HtmlFragmentSplitter.java]: I like this idea
`data-knotx-fragment-id` - should hold id of current fragment (new parameter)
`data-knotx-fallback` - should hold id of a other (fallback) fragment

File: knotx-core/src/main/java/io/knotx/assembler/FragmentAssemblerFallbackHandler.java [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+
+  public FragmentAssemblerFallbackHandler(FragmentAssemblerOptions options) {
+    this.options = options;
+    ServiceLoader<FallbackStrategy> fallbackStrategyServiceLoader = ServiceLoader.load(FallbackStrategy.class);
+    for (FallbackStrategy strategy : fallbackStrategyServiceLoader) {
+      fallbackStrategies.put(strategy.getId(), strategy);
+    }
+  }
+
+  public String applyFallback(Fragment failed, KnotContext knotContext) {
```
[the remark this answers is not in the document]
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/assembler/FragmentAssemblerFallbackHandler.java]: will implement lazy-initialized cache

File: knotx-core/src/main/java/io/knotx/dataobjects/KnotStatus.java [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.knotx.dataobjects;
+
+public enum KnotStatus {
```
[the remark this answers is not in the document]
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/dataobjects/KnotStatus.java]: will rename this class as well

**Comment by Skejven** [comment]
Please see my https://github.com/Cognifide/knotx/issues/466#issuecomment-435311790. I described there few implementation details that will make this solution easier to extend in the future.

File: knotx-core/src/main/java/io/knotx/assembler/FragmentAssemblerFallbackHandler.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/assembler/FragmentAssemblerFallbackHandler.java]: cache added to `FragmentAssemblerFallbackHandler`

File: knotx-core/src/main/java/io/knotx/splitter/HtmlFragmentSplitter.java (hunk 18 above)
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/splitter/HtmlFragmentSplitter.java]: I'm having some second thoughts on this, it would require additional capture groups in the regexps. I'm afraid we would take a performance hit here.

File: knotx-core/src/main/java/io/knotx/dataobjects/KnotStatus.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/dataobjects/KnotStatus.java]: class renamed

File: knotx-core/src/main/java/io/knotx/dataobjects/KnotError.java [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+import io.vertx.core.json.JsonObject;
+
+@DataObject(inheritConverter = true)
+public class KnotError {
+
+  private static final String CODE_KEY = "_CODE";
+  private static final String MESSAGE_KEY = "_MESSAGE";
+
+  private String code;
+  private Object message;
```
Comment (Skejven) [inline knotx-core/src/main/java/io/knotx/dataobjects/KnotError.java]: What is the reason for `message` to be the `Object` not `String`?

File: knotx-core/src/main/java/io/knotx/dataobjects/KnotTask.java [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+import com.google.common.base.Objects;
+import com.google.common.collect.Lists;
+import io.vertx.codegen.annotations.DataObject;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+import java.util.List;
+
+@DataObject(inheritConverter = true)
+public class KnotTask {
+  private String name;
```
Comment (Skejven) [inline knotx-core/src/main/java/io/knotx/dataobjects/KnotTask.java]: Shouldn't name be `final`?

File: knotx-core/src/main/java/io/knotx/assembler/FragmentAssemblerFallbackHandler.java [hunk 23]
```diff
@@ -0,0 +1,95 @@
...
+    FallbackStrategy strategy = getFallbackStrategy(fallback);
+    return strategy.applyFallback(failed, fallback, knotContext);
+  }
+
+  private FallbackStrategy getFallbackStrategy(Fragment fallbackFragment) {
+    String strategyId = Optional.ofNullable(fallbackFragment.getAttribute(FragmentConstants.FALLBACK_STRATEGY))
+        .orElse(DefaultFallbackStrategy.ID);
+    return Optional.ofNullable(fallbackStrategies.get(strategyId)).orElseThrow(() -> {
+      LOGGER.error("Fragment {} specifies fallback strategy but no fallback strategy with given id was found", fallbackFragment);
+      return new IllegalStateException(String.format("no strategy with id %s found", strategyId));
```
Comment (Skejven) [inline knotx-core/src/main/java/io/knotx/assembler/FragmentAssemblerFallbackHandler.java]: This is just cosmetics, but I'm not sure this is proper exception used here.
I see `IllegalStateException` as the wrong time to do some logic. Maybe `IllegalArgumentException` would be better fit here, what do you think?

File: knotx-core/src/main/java/io/knotx/options/SnippetOptions.java [hunk 24]
```diff
@@ -26,10 +29,15 @@
 public class SnippetOptions {
 
   private static final String DEFAULT_TAG_NAME = "script";
+  private static final String DEFAULT_FALLBACK_TAG_NAME = "knotx:fallback";
   private static final String DEFAULT_PARAMS_PREFIX = "data-knotx-";
+  private List<FallbackMetadata> DEFAULT_FALLBACKS = Lists.newArrayList(new FallbackMetadata("BLANK", "<knotx:fallback data-knotx-fallback-id=\"BLANK\"></knotx:fallback>"));
```
Comment (Skejven) [inline knotx-core/src/main/java/io/knotx/options/SnippetOptions.java]: Shouldn't the default content be:
`new FallbackMetadata("BLANK", "<"+tagName+":fallback "+paramsPrefix+"fallback-id=\"BLANK\"></"+tagName+":fallback>")`
?

Also, cosmetics: please fix the name of this field - it is not declared as a constant but has uppercase name.

File: knotx-core/src/main/java/io/knotx/splitter/HtmlFragmentSplitter.java (hunk 18 above)
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/splitter/HtmlFragmentSplitter.java]: Finally I decided to remove JSoup parsing from the Assembler. 
Additional regex to extract the fragment id is run only against fallback snippets.

File: knotx-core/src/main/java/io/knotx/dataobjects/KnotError.java (hunk 21 above)
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/dataobjects/KnotError.java]: I've had a request to make it an Object to have the ability to store structured data. Opting to keep it this way

File: knotx-core/src/main/java/io/knotx/dataobjects/KnotTask.java (hunk 22 above)
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/dataobjects/KnotTask.java]: Changed to final

File: knotx-core/src/main/java/io/knotx/assembler/FragmentAssemblerFallbackHandler.java (hunk 23 above)
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/assembler/FragmentAssemblerFallbackHandler.java]: I agree, changed to `IllegalArgumentException`

File: knotx-core/src/main/java/io/knotx/options/SnippetOptions.java (hunk 24 above)
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/options/SnippetOptions.java]: Method adjusted

File: knotx-core/src/main/java/io/knotx/assembler/FragmentAssemblerFallbackHandler.java [hunk 25]
```diff
@@ -0,0 +1,95 @@
...
+import org.apache.commons.lang3.StringUtils;
+
+public class FragmentAssemblerFallbackHandler {
+
+  private static final Logger LOGGER = LoggerFactory
+      .getLogger(FragmentAssemblerFallbackHandler.class);
+
+  private final Map<String, FallbackStrategy> fallbackStrategies = Maps.newHashMap();
+
+  private final Map<String, Fragment> fallbackFragmentCache = Maps.newHashMap();
```
[the remark this answers is not in the document]
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/assembler/FragmentAssemblerFallbackHandler.java]: I just pass new HashMap instance for each incoming request, I think this should do the trick?

File: documentation/src/main/wiki/FallbackHandling.md [hunk 26]
```diff
@@ -0,0 +1,134 @@
...
+To render a static markup instead of your failed snippet: 
+1. create a fallback snippet with `data-knotx-fallback-id` attribute 
+2. point to it in your snippet
+```
+<knotx-snippet data-knotx-knots="databridge,handlebars" 
+  data-knotx-databridge-name="unstable-service"
+  data-knotx-fallback="CUSTOM">
+      {{#if _result}}<h2>{{_result.count}}</h2>{{/if}}
+</knotx-snippet>
+<knotx:fallback data-knotx-fallback-id="CUSTOM">
```
Comment (Skejven) [inline documentation/src/main/wiki/FallbackHandling.md]: What do you think about changing `CUSTOM` to e.g. `MY_FALLBACK_ID`.
That might be more clear, that `CUSTOM` is not the mandatory id part for the custom fallback snippet.

File: knotx-core/src/main/java/io/knotx/knot/AbstractKnotProxy.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (Skejven) [reply knotx-core/src/main/java/io/knotx/knot/AbstractKnotProxy.java]: Let's mark it `@Deprecated` then.

Comment (Skejven) [inline knotx-core/src/main/java/io/knotx/knot/AbstractKnotProxy.java]: Let's mark `shouldProcess(Set<String> knots)` as `@Deprecated`. Method you suggested should replace it.

File: knotx-core/src/main/java/io/knotx/dataobjects/Fragment.java [hunk 27]
```diff
@@ -102,7 +145,45 @@ public JsonObject context() {
...
+  /**
+   * @return true if processing of this Fragment has failed
+   */
+  public boolean failed() {
+    return this.knots.stream().anyMatch(k -> k.getStatus() == KnotTaskStatus.FAILURE);
+  }
+
+  /**
+   * @return replacement markup that should be rendered if this Fragment has failed. Can be empty
+   * string. Null value indicates that no replacement markup is provided.
```
Comment (Skejven) [inline knotx-core/src/main/java/io/knotx/dataobjects/Fragment.java]: Could you please fix javadoc, this never should be null. You probably ment:
> absent value indicates that no replacement markup is provided.

Comment (scypio) [reply knotx-core/src/main/java/io/knotx/dataobjects/Fragment.java]: Fixed

File: documentation/src/main/wiki/FallbackHandling.md (hunk 26 above)
Comment (scypio) [reply documentation/src/main/wiki/FallbackHandling.md]: Changed

Comment (scypio) [reply documentation/src/main/wiki/FallbackHandling.md]: also: knotx-stack and knotx-example-project have PRs open that showcase this feature.

File: knotx-core/src/main/java/io/knotx/knot/AbstractKnotProxy.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (scypio) [reply knotx-core/src/main/java/io/knotx/knot/AbstractKnotProxy.java]: new task created to tackle this: https://github.com/Cognifide/knotx/issues/474

Comment (scypio) [reply knotx-core/src/main/java/io/knotx/knot/AbstractKnotProxy.java]: moved here: https://github.com/Cognifide/knotx/issues/474

File: knotx-core/src/main/java/io/knotx/assembler/FragmentAssemblerFallbackHandler.java [hunk 28]
```diff
@@ -0,0 +1,112 @@
...
+  private static final Logger LOGGER = LoggerFactory
+      .getLogger(FragmentAssemblerFallbackHandler.class);
+
+  private final Map<String, FallbackStrategy> fallbackStrategies = Maps.newHashMap();
+
+  private final FragmentAssemblerOptions options;
+
+  public FragmentAssemblerFallbackHandler(FragmentAssemblerOptions options) {
+    this.options = options;
+    ServiceLoader<FallbackStrategy> fallbackStrategyServiceLoader = ServiceLoader.load(FallbackStrategy.class);
```
[the remark this answers is not in the document]
Comment (Skejven) [reply knotx-core/src/main/java/io/knotx/assembler/FragmentAssemblerFallbackHandler.java]: Agreed to do this later.
